lib/system/arc.nim

```diff
@@ -202,15 +202,22 @@ proc nimDecRefIsLast(p: pointer): bool {.compilerRtl, inl.} =
         writeStackTrace()
         cfprintf(cstderr, "[DecRef] %p %ld\n", p, cell.count)
 
-    if cell.count == 0:
-      result = true
-      when traceCollector:
-        cprintf("[ABOUT TO DESTROY] %p\n", cell)
+    when defined(gcAtomicArc) and hasThreadSupport:
+      # `atomicDec` returns the new value
+      if atomicDec(cell.rc, rcIncrement) == -1:
+        result = true
+        when traceCollector:
+          cprintf("[ABOUT TO DESTROY] %p\n", cell)
     else:
-      decrement cell
-      # According to Lins it's correct to do nothing else here.
-      when traceCollector:
-        cprintf("[DECREF] %p\n", cell)
+      if cell.count == 0:
```

Comment:
Why? The old code was shorter and correct too.

Reply:
What do do you mean? I added a new branch for atomicArc.

Reply:
The changed code is for `atomicArc`

It is changed from
```nim
proc `=destroy`*(x: ref) =
  if atomicLoad(x.rc) == 0:
    `=destroy`(x[])
  else:
    atomicDec(x.rc)
```

```nim
proc `=destroy`*(x: ref) =
  if atomicDec(x.rc) == -1:
    `=destroy`(x[])
```

Reply:
My bad, I misread the diff.